File: 2900-2999/2996. Smallest Missing Integer Greater Than Sequential Prefix Sum/Solution.cpp
// Approach-1 (HashSet + Sequential Prefix)
// T.C : O(n) average
// S.C : O(n)

class Solution
{
public:
  int missingInteger(
      vector<int> &nums)
  {

    unordered_set<int> seen;

    for (
        int val : nums)
    {

      seen.insert(val);
    }

    int prefixSum =
        nums[0];

    for (
        int i = 1;
        i < nums.size();
        i++)
    {

      if (
          nums[i] ==
          nums[i - 1] + 1)
      {

        prefixSum +=
            nums[i];
      }
      else
      {

        break;
      }
    }

    while (
        seen.count(
            prefixSum))
    {

      prefixSum++;
    }

    return prefixSum;
  }
};

File: 2900-2999/2996. Smallest Missing Integer Greater Than Sequential Prefix Sum/Solution.java
//Approach-1 (HashSet + Sequential Prefix)
//T.C : O(n) average
//S.C : O(n)

import java.util.HashSet;
import java.util.Set;

class Solution {

  public int missingInteger(
      int[] nums) {

    Set<Integer> seen = new HashSet<>();

    for (int val : nums) {

      seen.add(val);
    }

    int prefixSum = nums[0];

    for (int i = 1; i < nums.length; i++) {

      if (nums[i] == nums[i - 1] + 1) {

        prefixSum += nums[i];

      } else {

        break;
      }
    }

    while (seen.contains(
        prefixSum)) {

      prefixSum++;
    }

    return prefixSum;
  }
}